src/System.IO.Pipelines/src/System/IO/Pipelines/SingleProducerSingleConsumerPool.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+using Internal;
+
+namespace System.IO.Pipelines
+{
+    internal sealed class SingleProducerSingleConsumerPool<T> where T : class
```

Comment:
The code in SingleProducerSingleConsumerQueue is pretty tricky.  Can you instead include/wrap that rather than borrowing the implementation?  That shared file could be modified if you have ways to improve it further, if you want to do minor refactorings to enable easier reuse of pieces, etc.  I'd like to avoid continuing to proliferate copies, though.

Reply:
Its unbounded for queuing currently (and multi-segment); whereas I'd want it bounded/single segment.

I can look to refactoring it into two parts and including the single segment? (as per ConcurrentQueue and ConcurrentQueueSegment)

Reply:
> I can look to refactoring it into two parts and including the single segment? 

Yeah, I'd prefer that approach.  Thanks!

Reply:
Might look at lock free in a future iteration